pkg/algorithm/nodeselect/v1alpha2/select_node.go

```diff
@@ -0,0 +1,206 @@
+}
+
+// getLabelSelectorString returns a string of label selector form label map to be used in
+// list options.
+func getLabelSelectorString(selector map[string]string) string {
+	var selectorString string
+	for key, value := range selector {
+		selectorString = selectorString + key + "=" + value + ","
+	}
+	selectorString = selectorString[:len(selectorString)-len(",")]
+	return selectorString
+}
+
+// GetUsedNode returns a map of node for which pool has already been created.
+func (ac *Config) GetUsedNode() (map[string]bool, error) {
+	usedNode := make(map[string]bool)
+	cspList, err := csp.
+		NewKubeClient().
+		WithNamespace(ac.Namespace).
+		List(
+			metav1.
+				ListOptions{LabelSelector: string(apis.CStorPoolClusterCPK) + "=" + ac.CSPC.Name},
+		)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not list already created csp(s)")
+	}
+	for _, cspObj := range cspList.Items {
+		usedNode[cspObj.Labels[string(apis.HostNameCPK)]] = true
+	}
+	return usedNode, nil
+}
+
+// GetBDListForNode returns a list of BD from the pool spec.
+// TODO : Move it to CStorPoolCluster packgage
+func (ac *Config) GetBDListForNode(pool *apis.PoolSpec) []string {
+	var BDList []string
+	for _, group := range pool.RaidGroups {
+		for _, bd := range group.BlockDevices {
+			BDList = append(BDList, bd.BlockDeviceName)
+		}
+	}
+	return BDList
+}
+
+// ClaimBDsForNode claims a given BlockDevice for node
+// If the block device(s) is/are already claimed for any other CSPC it returns error.
+// If the block device(s) is/are already calimed for the same CSPC -- it is left as it is and can be used for
+// pool provisioning.
+// If the block device(s) is/are unclaimed, then those are claimed.
+func (ac *Config) ClaimBDsForNode(BD []string) error {
+	for _, bdName := range BD {
+		bdAPIObj, err := bd.NewKubeClient().WithNamespace(ac.Namespace).Get(bdName, metav1.GetOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "error in getting details for BD {%s} whether it is claimed", bdName)
+		}
+		if bd.BuilderForAPIObject(bdAPIObj).BlockDevice.IsClaimed() {
+			IsClaimedBDUsable, err := ac.IsClaimedBDUsable(bdAPIObj)
+			if err != nil {
+				return errors.Wrapf(err, "error in getting details for BD {%s} for usability", bdName)
+			}
+			if !IsClaimedBDUsable {
+				return errors.Errorf("BD {%s} already in use", bdName)
+			}
+		}
+
+		err = ac.ClaimBD(bdAPIObj)
+		if err != nil {
+			return errors.Wrapf(err, "Failed to claim BD {%s}", bdName)
+		}
+	}
+	return nil
+}
+
+// ClaimBD claims a given BlockDevice
+func (ac *Config) ClaimBD(bdObj *ndmapis.BlockDevice) error {
+	newBDCObj, err := bdc.NewBuilder().
+		WithName("bdc-" + string(bdObj.UID)).
+		WithNamespace(ac.Namespace).
+		WithLabels(map[string]string{string(apis.CStorPoolClusterCPK): ac.CSPC.Name}).
+		WithBlockDeviceName(bdObj.Name).
+		WithHostName(bdObj.Labels[string(apis.HostNameCPK)]).
+		WithCapacity(volume.ByteCount(bdObj.Spec.Capacity.Storage)).
+		WithCSPCOwnerReference(ac.CSPC).
```

Comment:
can ownerReference itself if passed rather than CSPC object?

Reply:
Here -- I do not see a need but in deployment. Do you see CSP being owned by somebody else ?

Reply:
We will take it later.